File: apps/web/src/lib/api/presence.ts
/**
 * Owned-absence wire encoding (spec `lectern-player-lifecycle-hard-cutover.md`
 * §4): the one repository-wide forward encoding for a field whose absence is
 * a normal, successful outcome.
 *
 *   Presence<T> = { kind: "Absent" } | { kind: "Present"; value: T }
 *
 * The field is always present on the wire; `null`, omission, and alternate
 * casing are rejected. `decodePresence` is the same-system transport decoder
 * boundary (`docs/rules/boundaries.md`): decode once at the client boundary,
 * then pass the owned `Presence<T>` through model/view code unchanged. This
 * module is intentionally minimal.
 */

import { isRecord } from "@/lib/validation";

export type Presence<T> = { kind: "Absent" } | { kind: "Present"; value: T };

export function present<T>(value: T): Presence<T> {
  return { kind: "Present", value };
}

export function absent<T>(): Presence<T> {
  return { kind: "Absent" };
}

/**
 * Strictly decode a `Presence<T>` wire value. Rejects `null`/`undefined`,
 * non-object payloads, any `kind` other than the exact literals `"Absent"` or
 * `"Present"`, a `Present` missing `value`, and any extra keys on either
 * variant. `decodeValue` decodes the inner `value` for the `Present` case and
 * may itself throw to reject an invalid value.
 */
export function decodePresence<T>(
  raw: unknown,
  decodeValue: (value: unknown) => T,
): Presence<T> {
  if (!isRecord(raw)) {
    throw new Error(
      `Invalid Presence: expected an object, got ${raw === null ? "null" : typeof raw}`,
    );
  }
  const keys = Object.keys(raw);
  if (raw.kind === "Absent") {
    if (keys.length !== 1) {
      throw new Error(
        `Invalid Presence: "Absent" must have no keys besides "kind", got [${keys.join(", ")}]`,
      );
    }
    return { kind: "Absent" };
  }
  if (raw.kind === "Present") {
    if (keys.length !== 2 || !("value" in raw)) {
      throw new Error(
        `Invalid Presence: "Present" must have exactly "kind" and "value", got [${keys.join(", ")}]`,
      );
    }
    return { kind: "Present", value: decodeValue(raw.value) };
  }
  throw new Error(
    `Invalid Presence: "kind" must be "Absent" or "Present", got ${JSON.stringify(raw.kind)}`,
  );
}

/** Unwrap a decoded `Presence<T>` for view code that needs a plain fallback value. */
export function presenceValueOr<T>(presence: Presence<T>, fallback: T): T {
  return presence.kind === "Present" ? presence.value : fallback;
}

File: python/nexus/schemas/presence.py
"""Owned-absence wire encoding (spec `lectern-player-lifecycle-hard-cutover.md`
§4): the one repository-wide forward encoding for a field whose absence is a
normal, successful outcome.

```
Presence<T> = { kind: "Absent" } | { kind: "Present", value: T }
```

The field is always present on the wire; `null`, omission, and alternate
casing are rejected (the discriminator is a strict `Literal`, and every model
here is `extra="forbid"`). This module is intentionally minimal: the two
variant models, the generic alias used to annotate schema fields, two
internal constructors, and the two DB-adapter boundary converters named in
the spec. Nothing else belongs here — see `docs/rules/boundaries.md` for why
conversion helpers stay at the exact boundary that needs them.
"""

from __future__ import annotations

from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field


class Absent(BaseModel):
    """The `Presence<T>` absent variant. Carries no value."""

    kind: Literal["Absent"] = "Absent"

    model_config = ConfigDict(extra="forbid")


class Present[T](BaseModel):
    """The `Presence<T>` present variant. `value` is required."""

    kind: Literal["Present"] = "Present"
    value: T

    model_config = ConfigDict(extra="forbid")


# Generic alias usable directly as a field annotation, e.g. `Presence[int]`.
# PEP 695 syntax (not a plain `Annotated[...]` assignment) is required for the
# alias itself to be subscriptable; the discriminator is applied at every
# subscription site.
type Presence[T] = Annotated[Absent | Present[T], Field(discriminator="kind")]


def present[T](value: T) -> Present[T]:
    """Construct the present variant for internal code."""
    return Present[type(value)](value=value)


def absent() -> Absent:
    """Construct the absent variant for internal code."""
    return Absent()


def presence_from_nullable[T](value: T | None) -> Absent | Present[T]:
    """DB-adapter boundary only: convert a nullable DB column read into owned
    `Presence<T>`. Do not call this outside the query boundary that reads the
    nullable column (`docs/rules/boundaries.md`)."""
    if value is None:
        return Absent()
    return Present[type(value)](value=value)


def nullable_from_presence[T](value: Absent | Present[T]) -> T | None:
    """Final insert/update adapter only: convert owned `Presence<T>` back into
    the nullable form a DB column speaks. Do not call this outside the write
    boundary for that nullable column (`docs/rules/boundaries.md`)."""
    if isinstance(value, Absent):
        return None
    return value.value
